Sol.java

```diff
@@ -0,0 +1,104 @@
+import java.util.*;
+
+public class Sol {
+    int[] in ;
+    HashMap < Integer, Integer > exp;
+    LinkedList < Move > moves_coll;
+    int zeropos;
+
+    public LinkedList<Move> move_cars(int[] in , int[] out) {
+        initialise( in , out);
+        solve();
+      
+        return moves_coll;
+    }
+
+    private void initialise(int[] in , int[] out) {
+        exp = new HashMap < Integer, Integer > ();
+        moves_coll = new LinkedList < Move > ();
+        this.in = in ;
+
+        for (int i = 0; i < out.length; i++) {
+            exp.put(out[i], i);
+        }
+
+        for (int j = 0; j < in .length; j++)
+            if ( in [j] == 0) {
+                zeropos = j;
+                break;
+            }
+
+    }
+
+    private void solve() {
+        int i = 0;
+        while (i < in .length) {
+            if (exp.get( in [i]) != i && in [i] != 0) {
+                move(i);
+                i = -1;
```

Comment:
Because when I am finding the right place for a car X, if it's busy I will move Y to a free spot to free the current spot and put the other car X. That free spot can be at the beginning so I need to go there and take that car Y and put it where it should be.